rpcserver.go

```diff
@@ -1875,6 +1880,337 @@ func (r *rpcServer) AnchorVirtualPsbts(ctx context.Context,
+			continue
+		}
+
+		// If we already have the derivation info, we can skip this
+		// input.
+		if len(pIn.TaprootBip32Derivation) > 0 {
+			continue
+		}
+
+		// Let's query our node for the internal key information now.
+		internalKey, keyLocator, err := r.querySchnorrInternalKey(
+			ctx, pIn.TaprootInternalKey,
+		)
+
+		switch {
+		case errors.Is(err, address.ErrInternalKeyNotFound):
+			// If the internal key is not known, we can't add the
+			// derivation info. Most likely this asset input is not
+			// owned by this daemon but another party.
+			continue
+
+		case err != nil:
+			return nil, fmt.Errorf("error querying internal key: "+
+				"%w", err)
+		}
+
+		keyDesc := keychain.KeyDescriptor{
+			PubKey:     internalKey,
+			KeyLocator: keyLocator,
+		}
+		derivation, trDerivation := tappsbt.Bip32DerivationFromKeyDesc(
+			keyDesc, r.cfg.ChainParams.HDCoinType,
+		)
+		pIn.Bip32Derivation = []*psbt.Bip32Derivation{derivation}
+		pIn.TaprootBip32Derivation = []*psbt.TaprootBip32Derivation{
+			trDerivation,
+		}
+	}
+
+	// TODO(guggero): Validate inputs/outputs by going through all vPSBTs.
+
+	// We're ready to attempt to fund the transaction now. For that we first
+	// need to re-serialize our packet.
+	var buf bytes.Buffer
+	err = pkt.Serialize(&buf)
+	if err != nil {
+		return nil, fmt.Errorf("error serializing packet: %w", err)
+	}
+
+	// The change output and fee parameters of this RPC are identical to the
+	// walletrpc.FundPsbt, so we just map them 1:1 and let lnd do the
+	// validation.
+	coinSelect := &walletrpc.PsbtCoinSelect{
+		Psbt: buf.Bytes(),
+	}
+	fundRequest := &walletrpc.FundPsbtRequest{
+		Template: &walletrpc.FundPsbtRequest_CoinSelect{
+			CoinSelect: coinSelect,
+		},
+		MinConfs:   1,
+		ChangeType: P2TrChangeType,
+	}
+
+	// Unfortunately we can't use the same RPC types, so we have to do a
+	// 1:1 mapping to the walletrpc types for the anchor change output and
+	// fee "oneof" fields.
+	type existingIndex = walletrpc.PsbtCoinSelect_ExistingOutputIndex
+	switch change := req.AnchorChangeOutput.(type) {
+	case *wrpc.CommitVirtualPsbtsRequest_ExistingOutputIndex:
+		coinSelect.ChangeOutput = &existingIndex{
+			ExistingOutputIndex: change.ExistingOutputIndex,
+		}
+
+	case *wrpc.CommitVirtualPsbtsRequest_Add:
+		coinSelect.ChangeOutput = &walletrpc.PsbtCoinSelect_Add{
+			Add: change.Add,
+		}
+
+	default:
+		return nil, fmt.Errorf("unknown change output type")
+	}
+
+	switch fee := req.Fees.(type) {
+	case *wrpc.CommitVirtualPsbtsRequest_TargetConf:
+		fundRequest.Fees = &walletrpc.FundPsbtRequest_TargetConf{
+			TargetConf: fee.TargetConf,
+		}
+
+	case *wrpc.CommitVirtualPsbtsRequest_SatPerVbyte:
+		fundRequest.Fees = &walletrpc.FundPsbtRequest_SatPerVbyte{
+			SatPerVbyte: fee.SatPerVbyte,
+		}
+
+	default:
+		return nil, fmt.Errorf("unknown fee type")
+	}
+
+	lndWallet := r.cfg.Lnd.WalletKit
+	fundedPacket, changeIndex, lockedUTXO, err := lndWallet.FundPsbt(
```

Comment:
IIUC at this point there is a strict check that the inputs amount must equal the outputs amount.

If I want to create an incomplete anchorTx, could we somehow signal that there may be an input or output missing, hence no need to perform strict checks?

Reply:
No, the input and output sum doesn't have to be equal here. That's exactly the task for `lndWallet.FundPsbt` to figure out if additional inputs are required to fund the outputs.
But that does mean that all _outputs_ need to be specified here, otherwise we wouldn't know what to fund.